src/EditorFeatures/Core/Implementation/Diagnostics/DiagnosticsSquiggleTaggerProvider.TagSource.cs

```diff
@@ -59,27 +60,60 @@ protected override TagSpan<IErrorTag> CreateTagSpan(SnapshotSpan span, Diagnosti
                 return new TagSpan<IErrorTag>(span, new ErrorTag(errorType, diagnostic.Message));
             }
 
-            private string GetErrorTypeFromDiagnosticKind(DiagnosticData diagnostic)
+            private string GetErrorTypeFromDiagnostic(DiagnosticData diagnostic)
             {
-                if (diagnostic.CustomTags.Count > 1)
+                var fromTag = GetErrorTypeFromDiagnosticTags(diagnostic);
+                if (fromTag != null)
                 {
-                    switch (diagnostic.CustomTags[0])
-                    {
-                        case WellKnownDiagnosticTags.EditAndContinue:
-                            return EditAndContinueErrorTypeDefinition.Name;
-                        case WellKnownDiagnosticTags.Build:
-                            if (_blueSquiggleForBuildDiagnostic)
-                            {
-                                return PredefinedErrorTypeNames.CompilerError;
-                            }
-
-                            break;
-                    }
+                    return fromTag;
+                }
+
+                var fromProperty = GetErrorTypeFromDiagnosticProperty(diagnostic);
+                if (fromProperty != null)
+                {
+                    return fromProperty;
                 }
 
                 return GetErrorTypeFromDiagnosticSeverity(diagnostic);
             }
 
+            private string GetErrorTypeFromDiagnosticProperty(DiagnosticData diagnostic)
+            {
+                if (diagnostic.Properties.Count == 0)
+                {
+                    return null;
+                }
+
+                string value;
+                if (!diagnostic.Properties.TryGetValue(WellKnownDiagnosticPropertyNames.Origin, out value))
+                {
+                    return null;
+                }
+
+                if (value == WellKnownDiagnosticTags.Build && _blueSquiggleForBuildDiagnostic)
+                {
+                    return PredefinedErrorTypeNames.CompilerError;
+                }
+
+                return null;
+            }
+
+            private string GetErrorTypeFromDiagnosticTags(DiagnosticData diagnostic)
+            {
+                if (diagnostic.CustomTags.Count <= 1)
+                {
+                    return null;
+                }
+
+                switch (diagnostic.CustomTags[0])
```

Comment:
there is unit test to check it is the first one. once we support adding custom tagger for diagnostic, we probably had to search but since it is open-range, I am worrying about perf.